src/tree/updater_quantile_hist.cc

```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
       this->EvaluateSplit(cright, gmat, hist_, *p_fmat, *p_tree);
       time_evaluate_split += dmlc::GetTime() - tstart;
 
-      qexpand_->push(ExpandEntry(cleft, p_tree->GetDepth(cleft),
+      qexpand_loss_guided_->push(ExpandEntry(cleft, p_tree->GetDepth(cleft),
                                  snode_[cleft].best.loss_chg,
                                  timestamp++));
-      qexpand_->push(ExpandEntry(cright, p_tree->GetDepth(cright),
+      qexpand_loss_guided_->push(ExpandEntry(cright, p_tree->GetDepth(cright),
                                  snode_[cright].best.loss_chg,
                                  timestamp++));
 
       ++num_leaves;  // give two and take one, as parent is no longer a leaf
     }
   }
+}
+
+void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+                                    const GHistIndexBlockMatrix& gmatb,
+                                    const ColumnMatrix& column_matrix,
+                                    HostDeviceVector<GradientPair>* gpair,
+                                    DMatrix* p_fmat,
+                                    RegTree* p_tree) {
+  double gstart = dmlc::GetTime();
+
+  const std::vector<GradientPair>& gpair_h = gpair->ConstHostVector();
+
+  spliteval_->Reset();
+
+  tstart = dmlc::GetTime();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  time_init_data = dmlc::GetTime() - tstart;
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1)
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+    while (!qexpand_loss_guided_->empty()) {
+      const int nid = qexpand_loss_guided_->top().nid;
+      qexpand_loss_guided_->pop();
+      (*p_tree)[nid].SetLeaf(snode_[nid].weight * param_.learning_rate);
+    }
+  } else {
+    ExpandWithDepthWidth(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+  }
 
   // set all the rest expanding nodes to leaf
```

Comment:
Is this comment still valid?

Reply:
This comment should be moved to line 359

Reply:
NVM, let's just remove the comment